marconi/src/Marconi/Index/New/ScriptTx.hs

```diff
@@ -0,0 +1,418 @@
+-- import Control.Applicative (Alternative (some))
+-- import Control.Monad (when)
+import RewindableIndex.Index.VSqlite (SqliteIndex)
+-- import RewindableIndex.Index.VSqlite qualified as Ix
+import RewindableIndex.Storable (Buffered (getStoredEvents, persistToStorage), HasPoint (getPoint),
+                                 QueryInterval (QEverything, QInterval), Queryable (queryStorage),
+                                 Rewindable (rewindStorage), StorableEvent, StorableMonad, StorablePoint, StorableQuery,
+                                 StorableResult, emptyState, filterWithQueryInterval)
+import RewindableIndex.Storable qualified as Storable
+-- import GHC.Exts (the)
+
+{- The first thing that we need to define for a new indexer is the `handler` data
+   type, meant as a wrapper for the connection type (in this case the SQLite
+   connection).
+
+   However this is a very good place to add some more configurations
+   that the indexer may require (for example performance tuning settings). In our
+   case we add the number of events that we want to return from the on-disk buffer -}
+
+data ScriptTxHandle = ScriptTxHandle
+  { hdlConnection :: SQL.Connection
+  , hdlDiskStore  :: Int
+  }
+
+{- The next step is to define the data types that make up the indexer. There are
+   5 of these and they depend on the handle that we previously defined. We make use
+   of this semantic dependency by using type and data families that connect these
+   types to the `handle` that was previously defined.
+
+   If you want to think of semantics, you can thing of the `handle` type as identifying
+   both the database connection type and the database structure. Thinking of it this
+   way makes the reason for the dependency clearer.
+
+   The first type we introduce is the monad in which the database (and by extension,
+   the indexer) runs. -}
+
+type instance StorableMonad ScriptTxHandle = IO
+
+{- The next type we introduce is the type of events. Events are the data atoms that
+   the indexer consumes. They depend on the `handle` because they need to eventually
+   be persisted in the database, so the database has to be able to accomodate them.
+
+   THe original implementation used two spearate data structures for storing data
+   in memory vs. on-disk. It has the advantage of a better usage of memory and the
+   disadvantage of complicating the implementation quite a bit. I am leaving it as-is
+   for now, as this is more of a tutorial implementation and complicating things
+   may have some educational value. -}
+
+data instance StorableEvent ScriptTxHandle = ScriptTxEvent
+  { txScripts :: [(TxCbor, [StorableQuery ScriptTxHandle])]
+  , slotNo    :: !SlotNo
+  } deriving (Show)
+
+{- The resume and query functionality requires a way to specify points on the chain
+   from which we want to resume, or points up to which we want to query. Next we
+   define the types of these points. -}
+
+type instance StorablePoint ScriptTxHandle = SlotNo
```

Comment:
Curious: isn't it boilerplate because the resume/query point is **always** of type `SlotNo`? Is there reason to have another datatype? (Other than maybe the fact someone might use different types than the ones from `cardano-api`)

Reply:
Thanks :)

Reply:
Also, think about side-chains (midnight). Or proprietary data streams coming from centralised exchanges (there was a project for building a dex aggregator that I am thinking about). Maybe one would want to correlate them somehow in a hybrid measure of time. In that case we may see very different types of points.